types/node/fs.d.ts

```diff
@@ -4332,15 +4344,26 @@ declare module "fs" {
         pattern: string | string[],
         callback: (err: NodeJS.ErrnoException | null, matches: string[]) => void,
     ): void;
-    export function glob(
+    export function glob<Options extends GlobOptions>(
         pattern: string | string[],
-        options: GlobOptions,
-        callback: (err: NodeJS.ErrnoException | null, matches: string[]) => void,
+        options: Options,
+        callback: (
+            err: NodeJS.ErrnoException | null,
+            matches: Options extends GlobOptionsWithFileTypes ? Dirent[]
+                : Options extends GlobOptionsWithoutFileTypes ? string[]
+                : Dirent[] | string[],
+        ) => void,
     ): void;
```

Comment:
Looking around, other methods with option-dependent types (crypto.generatePrime, dns.lookup, etc.) use overload definitions, rather than conditional types. This might be a cleaner approach.

```ts
    export function glob(
        pattern: string | string[],
        options: GlobOptionsWithFileTypes,
        callback: (
            err: NodeJS.ErrnoException | null,
            matches: Dirent[],
        ) => void,
    ): void;
    export function glob(
        pattern: string | string[],
        options: GlobOptionsWithoutFileTypes,
        callback: (
            err: NodeJS.ErrnoException | null,
            matches: string[],
        ) => void,
    ): void;
    export function glob(
        pattern: string | string[],
        options: GlobOptions,
        callback: (
            err: NodeJS.ErrnoException | null,
            matches: Dirent[] | string[],
        ) => void,
    ): void;
```